File: leet/1492_kth_factor_of_n/SolutionB.java
// LeetCode 1492. The kth Factor of n
// O(root(n)) time, O(1) space
class Solution {
    public int kthFactor(int n, int k) {
        double root = Math.sqrt(n);
        for (int i = 1; i < root; i++)
            if (n % i == 0 && --k == 0)
                return i;
        for (int i = (int)root; i > 0; i--)
            if (n % i == 0 && --k == 0)
                return n / i;
        return -1;
    }
}

File: leet/1463_cherry_pickup_2/solution_b.py
# LeetCode 1463. Cherry Pickup II
# O(m*n^2) time & space
class Solution:
    def cherryPickup(self, grid: List[List[int]]) -> int:
        @lru_cache(None)
        def dp(i, j, k):
            if not 0 <= j < len(grid[0]) or not 0 <= k < len(grid[0]) or j == k:
                return float('-inf')
            
            result = grid[i][j] + grid[i][k]
            if i == len(grid) - 1:
                return result
            _max = float('-inf')
            for j2 in range(j - 1, j + 2):
                for k2 in range(k - 1, k + 2):
                    _max = max(_max, dp(i + 1, j2, k2))
            return result + _max

        return dp(0, 0, len(grid[0]) - 1)
